### PR TITLE
Dmay/job launcher refactor

### PR DESCRIPTION
**Pull request by psanan**

### CODE REVIEW

**Comment by psanan** [comment]
I want to make sure that the existing examples (which are part of the current tests) continue to work. Here's a proposed plan to do that:

1. Add a `VerifierLegacy` class (in its own file, to be removed once obsolete)
2. Move all the `compareXYZ()` functions which to be class methods of `VerifierLegacy`
3. Update the examples, hence tests, to use this (expecting to update them to use new-and-improved approaches later). If some example is particularly laborious to update this way, I might just remove it.

[Update: not doing this, too much pointless work]

**Comment by psanan** [comment]
I think it's a waste of time to make the existing harness work, before refactoring it, so I'll restrict my ambitions here to simply add the new tests to the CI, taking the tests of the existing examples out for now, to be re-added as the Harness is refactored.

**Comment by dmay23** [comment]
Yes! Delete harness and start again. And thus disable old tests for the
time being. That’s what we agreed upon when we spoke last week. When I
reimplement smartDiff I’ll put the old (non harness) tests back in.

File: sciath/launcher.py [hunk 1]
```diff
@@ -48,135 +48,268 @@ def formatMPILaunchCommand(mpiLaunch,ranks,corespernode):
...
+            jprefix = "sciath.job-" + job.name
+        
+        stdoutName.append( jprefix + ".stdout" )
+        stderrName.append( jprefix + ".stderr" )
+
+        lc_count -= 1
+
+    return errorCodeName, stdoutName, stderrName
+
+def _removeFile(file2rm):
```
Comment (psanan) [inline sciath/launcher.py:79]: Is it safe to turn off "safety mode" here? I am trying to add your new tests to the self-tests before proceeding, but they need to run without user input.

Comment (psanan) [reply sciath/launcher.py:79]: (For now, working around this by doing e.g. `yes | ./runTests.py`)

**Comment by dmay23** [comment]
Yes, safe to turn this off. It’s primarily intended for safety when
developing debugging the harness / sandbox-ing where there is lots of path
manipulations.

Maybe make safety mode a global module variable. As per any other debugging
flags. They are very useful and avoid trashing something by mistake.

**Comment by psanan** [comment]
@dmay23 I'd like to merge this, since it's already huge, and work on remaining fixes and improvements in new PRs to dev.

Note that this PR totally breaks compatibility with pyTestHarness, but I'm not bothering to bump the version since the other open PR does that.

**Comment by dmay23** [comment]
Why did you use os.system()? Use the internal py2/3 compatible function
please.

**Comment by psanan** [comment]
Because I tried it and it failed, and you had it commented out so I figured it wasn't trivial, and I wanted something running that we could iterate on in other, contained, issues (opened issue #50 for this). Is it going to work when the command has a redirect in it? (e.g. `bsub < foo.lsf`).
```
Created submission file: ./sandbox/sciath.job-DMDA_interpolation-launch.lsf
[Executing DMDA_interpolation]
  [cmd]  ['bsub < ', './sandbox/sciath.job-DMDA_interpolation-launch.lsf']
Traceback (most recent call last):
  File "./test3.py", line 19, in <module>
    job_launcher.submitJob(jA,path=path)
  File "/cluster/home/pasanan/code/sciath/sciath/launcher.py", line 652, in submitJob
    ctx = subp.run(launchCmd)
  File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 403, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 707, in __init__
    restore_signals, start_new_session)
  File "/cluster/apps/python/3.6.0/x86_64/lib64/python3.6/subprocess.py", line 1326, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'bsub < '
```

**Comment by dmay23** [comment]
Sure sure, but let’s not put the code which is lame and which we want
removed back in - rather let’s get the improved approach working (note I’ve
removed os.system() completely from Test, Job, Launcher).

Recall I said “code not at all tested on any batch system”.  So I never
tested the commented line either. :(

Can you try this:
cmd = [‘bsub’, ’<‘,  ’foo.lsf’]
In general there should be no white space.

**Comment by psanan** [comment]
I don't think that works directly
```
$ python
Python 3.7.4 (default, Sep  7 2019, 19:52:29)
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> subprocess.run(["cat","test.txt"])
foo
CompletedProcess(args=['cat', 'test.txt'], returncode=0)
>>> subprocess.run(["cat","<","test.txt"])
cat: <: No such file or directory
foo
CompletedProcess(args=['cat', '<', 'test.txt'], returncode=1)
```
(Also see e.g. https://stackoverflow.com/questions/3679974/run-shell-command-with-input-redirections-from-python-2-4)

**Comment by dmay23** [comment]
Ok. So read the file contents and shove it into stdin (see second answer)

**Comment by psanan** [comment]
What do you think about this idea? We add some more logic to our custom `run()` function, so that in addition to working with python 2 and 3, it can (optionally, on by default) also interpret some small subset of the bash redirection symbols. In particular (and this is all we'd initially implement), it would look for an entry in the command list equal to `"<"`, in the second-to-last position, and if found, use the last argument to open a file and send it as stdin to a function call corresponding to the remaining arguments. This would cover the `bsub <` case (which would become the `['bsub','<']` case).

**Comment by psanan** [comment]
To be more concrete, and for reference when I come back to this, I mean something like this:
```
def run(command,stdoutFile,stderrFile,interpret_redirect_symbols=True) :
  """ Command akin to subprocess.run(), but with both Python 2 and Python 3
      support, and with an additional option to specially interpret
      a unix-style redirect "<" in the second-to-last slot """
  if stdoutFile is None:
      stdoutFile = subp.PIPE
  if stderrFile is None:
      stderrFile = subp.PIPE

  if interpret_redirect_symbols and len(command) >= 3 and command[-2] == '<':
      stdinFile = open(command[-1],'r')
      command = command[:-2]
  else:
      stdinFile = None

  try:
      if sys.version_info[0] == 2 :
          errorno = subp.call(command,stdin=stdinFile,stdout=stdoutFile,stderr=stderrFile)
      else :
          ctx = subp.run(command,universal_newlines=True,stdin=stdinFile,stdout=stdoutFile,stderr=stderrFile)
          errorno = ctx.returncode
  finally:
      if stdinFile:
          stdinFile.close()

  return errorno
```

**Comment by dmay23** [comment]
On Tue, 5 Nov 2019 at 07:28, Patrick Sanan <notifications@github.com> wrote:

> What do you think about this idea? We add some more logic to our custom
> run() function, so that in addition to working with python 2 and 3, it
> can (optionally, on by default) also interpret some small subset of the
> bash redirection symbols. In particular (and this is all we'd initially
> implement), it would look for an entry in the command list equal to "<",
> in the second-to-last position, and if found, use the last argument to open
> a file and send it as stdin to a function call corresponding to the
> remaining arguments. This would cover the bsub < case (which would become
> the ['bsub','<'] case).
>

Sounds okay to me.... but....
There is only one line in the entire code base which needs < .
So rather do something complex like changing the behaviour of what happens
when the second last item in the list of commands was "<", for this one
line in sciath where < is required I would preferably opt to just open the
file and jam it in to subp as stdin...

**Comment by psanan** [comment]
It's not precisely "one line"  - the launcher will somehow have to store information about this redirect, which is annoying in the first place because it breaks the convenience assumption that a "command" is a list of strings [exec, arg, arg, .. ].  I'm not sure that handling this will be any less complexity! (Though you can argue that this approach could cause confusion in that we only handle a subset of things you can do in bash - I could just add some error messages for now if you have a ">", "|", or a "<" in any position but the second-to-last).


> Am 07.11.2019 um 12:31 schrieb dmay23 <notifications@github.com>:
>

**Comment by psanan** [comment]
Pushed what I think is a nice, non-invasive option. Instead of `bsub <`, which `subprocess.run()` can't handle because it doesn't interpret `<`, just pass through `sh` which can run a string as commands, handling the redirects, and accept arguments, i.e. `sh -c 'bsub < $0'` (Note single quotes). This seems to work fine and keeps everything simple.

**Pull request merged**